src/result.alp

```diff
@@ -0,0 +1,57 @@
+{- Basic result type.
+ -}
+module result
+
+import_type option.option
+
+export_type result
+
+type result 'a 'b = Ok 'a | Err 'b
+
+let is_ok r =
+  match r with
+    Ok _ -> true
+  | Err _ -> false
+
+let is_ok r =
+  match r with
+    Ok _ -> false
+  | Err _ -> true
+
+test "Applying `is_ok` to Ok x should return true" =
+  assert.equal (true) (is_ok (Ok 1))
+
+let map f (Ok x) = Ok (f x)
+
+let map _ (Err x) = Err x
+
+test "mapping the identity function to Ok x should return Ok x" =
+  let f x = x in
+  assert.equal (Ok 1) (map f (Ok 1))
+
+let of_option opt e =
+  match opt with
+    Some x -> Ok x
+  | None -> Err e
+
+let andThen callback (Ok value) = callback value
```

Comment:
Here we are using camelCase while everywhere else we are using snake_case. Could we be consistent?

Reply:
Have we decided which case we prefer? I'm leaning towards camelCase.

Reply:
iirc @lepoetemaudit and I discussed this mid-way through last year and settled on moving to camelCase.  I'm easy either way tbh.  Other opinions?

Reply:
Fab, I'll update the other modules/PR.

Reply:
For option this is called `flatMap`. Which name do we prefer? I took `flatMap` from Scala as we had `option` which felt Scala-y to me.

Reply:
I prefer `andThen` personally, but would also change it in `option`, hehe. So if it is staying that way in option then to be consistent I'll change it here.

Maybe in scala the option and result types are actually related to sequences and so a `flatMap` makes sense? To me it doesn't fit in the alpaca case and would be confusing.

Reply:
It's the result of composing `flatten : m m a -> m a` and `map`, so I think it makes sense.

I find the name `andThen` a little odd, doesn't seem to fit with `map`, `fold` etc